tenseal/tensors/ckksvector.cpp

```diff
@@ -412,4 +415,60 @@ CKKSVector& CKKSVector::replicate_first_slot_inplace(size_t n) {
     return *this;
 }
 
+CKKSVector CKKSVector::polyval(const vector<double>& coefficients) {
+    CKKSVector new_vector = *this;
+    return new_vector.polyval_inplace(coefficients);
+}
+
+CKKSVector& CKKSVector::polyval_inplace(const vector<double>& coefficients) {
+    if (coefficients.size() == 0) {
+        throw invalid_argument(
+            "the coefficients vector need to have at least one element");
+    }
+
+    int degree = coefficients.size() - 1;
+    while (degree >= 0) {
+        if (coefficients[degree] == 0.0)
+            degree--;
+        else
+            break;
+    }
+
+    // null polynomial: output should be an encrypted 0
+    // we can multiply by 0, or return the encryption of zero
+    if (degree == -1) {
+        // we set the vector to the encryption of zero
+        this->context->encryptor->encrypt_zero(this->ciphertext);
+        this->ciphertext.scale() = this->init_scale;
+        return *this;
+    }
+
+    // set result accumulator to the constant coefficient
+    vector<double> cst_coeff(this->size(), coefficients[0]);
+    CKKSVector result(this->context, cst_coeff, this->init_scale);
+
+    // pre-compute squares of x
+    CKKSVector x = *this;
+    int max_square = static_cast<int>(floor(log2(degree)));
+    cout << "max square " << max_square << endl;
+    vector<CKKSVector> x_squares;
+    x_squares.reserve(max_square + 1);
```

Comment:
Why not directly resize? and access the vector directly using [] instead of push_back operations

Reply:
True!

Reply:
**Update:** So it's a little complicated for this, reserve is more convenient for our usage, as resize need to init the objects in the vector, and we doesn't want this, however, reserve will only allocate the necessary memory and push_back won't have to allocate memory dynamically.

Reply:
Yeah, my observation made more sense with vector of pointers.
Sorry about that.

Reply:
Thank you!